src/gsad.c

```diff
@@ -3828,6 +3833,120 @@ gsad_add_content_type_header (struct MHD_Response *response,
     }
 }
 
+/**
+ * @brief Add all local IP addresses to a GHashTable.
+ *
+ * @param[in] hashtable     The hashtable to add the addresses to.
+ * @param[in] include_ipv6  Whether to include IPv6 addresses.
+ */
+void
+add_local_addresses (GHashTable *hashtable, int include_ipv6)
+{
+  struct ifaddrs *ifaddr, *ifa;
+  int family, ret;
+  char host[NI_MAXHOST];
+
+  if (getifaddrs(&ifaddr) != -1)
+    {
+      for (ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next)
+        {
+          if (ifa->ifa_addr == NULL)
+            continue;
+
+          family = ifa->ifa_addr->sa_family;
+
+          if (family == AF_INET || (include_ipv6 && family == AF_INET6))
+            {
+              ret = getnameinfo(ifa->ifa_addr,
+                                (family == AF_INET) 
+                                  ? sizeof(struct sockaddr_in) 
+                                  : sizeof(struct sockaddr_in6),
+                                host, NI_MAXHOST,
+                                NULL, 0, NI_NUMERICHOST);
+              if (ret != 0) {
+                g_warning ("%s: getnameinfo() failed: %s\n",
+                           __FUNCTION__, gai_strerror(ret));
+                return;
+              }
+
+              g_hash_table_insert (hashtable, g_strdup (host), NULL);
+            }
+        }
+      freeifaddrs(ifaddr);
+    }
+}
+
+/**
+ * @brief Verify if a hostname or IP address refers to the gsad.
+ *
+ * @param[in] host  The name or address to check.
+ *
+ * @return 0 valid, 1 invalid.
+ */
+static int
+host_is_gsad (const char *host)
+{
+  if (host == NULL)
+    return 0;
+
+  return host ? g_hash_table_contains (gsad_header_hosts, host) : 0;
+}
+
+/**
+ * Verify a Host HTTP header.
+ *
+ * @param[in] host_header  Host header value.
+ *
+ * @return 0 valid, 1 invalid UTF-8, 2 otherwise invalid, -1 error.
+ */
+static int
+validate_host_header (const char *host_header)
+{
+  int ret;
+  int char_index, colon_index, bracket_index;
+  gchar *host;
+
+  if (host_header == NULL || strlen (host_header) == 0)
+    return 2;
+  else if (g_utf8_validate (host_header, -1, NULL) == FALSE)
+    return 1;
+
+  bracket_index = -1;
+  colon_index = -1;
+  for (char_index = strlen (host_header) - 1;
```

Comment:
I don't really understand this code. Maybe you can extract it into another function or at least add some comment.